core/src/cluster_nodes.rs

```diff
@@ -0,0 +1,441 @@
+    itertools::Itertools,
+    solana_gossip::{
+        cluster_info::{compute_retransmit_peers, ClusterInfo},
+        contact_info::ContactInfo,
+        crds_gossip_pull::CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS,
+        weighted_shuffle::{weighted_best, weighted_shuffle},
+    },
+    solana_sdk::pubkey::Pubkey,
+    std::{any::TypeId, cmp::Reverse, collections::HashMap, marker::PhantomData},
+};
+
+enum NodeId {
+    // TVU node obtained through gossip (staked or not).
+    ContactInfo(ContactInfo),
+    // Staked node with no contact-info in gossip table.
+    Pubkey(Pubkey),
+}
+
+struct Node {
+    node: NodeId,
+    stake: u64,
+}
+
+pub struct ClusterNodes<T> {
+    pubkey: Pubkey, // The local node itself.
+    // All staked nodes + other known tvu-peers + the node itself;
+    // sorted by (stake, pubkey) in descending order.
+    nodes: Vec<Node>,
+    // Weights and indices for sampling peers. weighted_{shuffle,best} expect
+    // weights >= 1. For backward compatibility we use max(1, stake) for
+    // weights and exclude nodes with no contact-info.
+    index: Vec<(/*weight:*/ u64, /*index:*/ usize)>,
+    _phantom: PhantomData<T>,
+}
+
+impl Node {
+    #[inline]
+    fn pubkey(&self) -> Pubkey {
+        match &self.node {
+            NodeId::Pubkey(pubkey) => *pubkey,
+            NodeId::ContactInfo(node) => node.id,
+        }
+    }
+
+    #[inline]
+    fn contact_info(&self) -> Option<&ContactInfo> {
+        match &self.node {
+            NodeId::Pubkey(_) => None,
+            NodeId::ContactInfo(node) => Some(node),
+        }
+    }
+}
+
+impl<T> ClusterNodes<T> {
+    pub fn num_peers(&self) -> usize {
+        self.index.len()
+    }
+
+    // A peer is considered live if they generated their contact info recently.
+    pub fn num_peers_live(&self, now: u64) -> usize {
+        self.index
+            .iter()
+            .filter_map(|(_, index)| self.nodes[*index].contact_info())
+            .filter(|node| {
+                let elapsed = if node.wallclock < now {
+                    now - node.wallclock
+                } else {
+                    node.wallclock - now
+                };
+                elapsed < CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS
+            })
+            .count()
+    }
+}
+
+impl ClusterNodes<BroadcastStage> {
+    pub fn new(cluster_info: &ClusterInfo, stakes: &HashMap<Pubkey, u64>) -> Self {
+        new_cluster_nodes(cluster_info, stakes)
+    }
+
+    /// Returns the root of turbine broadcast tree, which the leader sends the
+    /// shred to.
+    pub fn get_broadcast_peer(&self, shred_seed: [u8; 32]) -> Option<&ContactInfo> {
+        if self.index.is_empty() {
+            None
+        } else {
+            let index = weighted_best(&self.index, shred_seed);
+            match &self.nodes[index].node {
+                NodeId::ContactInfo(node) => Some(node),
+                NodeId::Pubkey(_) => panic!("this should not happen!"),
+            }
+        }
+    }
+}
```

Comment:
not to self: should be equivalent to deleted code: https://github.com/solana-labs/solana/blob/master/core/src/broadcast_stage.rs#L421-L423

Reply:
tag @jbiseda, would be awesome if you could double check this as well!

Reply:
yes, this is tested in `test_cluster_nodes_broadcast`